src/dfx/src/lib/environment.rs

```diff
@@ -243,24 +244,41 @@ pub struct AgentEnvironment<'a> {
     backend: &'a dyn Environment,
     agent: Agent,
     network_descriptor: NetworkDescriptor,
-    identity_manager: IdentityManager,
+    identity_manager: Option<IdentityManager>,
 }
 
 impl<'a> AgentEnvironment<'a> {
-    #[context("Failed to create AgentEnvironment for network '{}'.", network_descriptor.name)]
     pub fn new(
         backend: &'a dyn Environment,
         network_descriptor: NetworkDescriptor,
         timeout: Duration,
     ) -> DfxResult<Self> {
-        let logger = backend.get_logger().clone();
         let mut identity_manager = IdentityManager::new(backend)?;
         let identity = identity_manager.instantiate_selected_identity(backend.get_logger())?;
```

Comment:
Suggestion: Reduce special-casing for the anonymous identity by adding a parameter `use_identity: Option<String>` here, and then either calling `identity_manager.instantiate_selected_identity` or `identity_manager.instantiate_identity_from_name`.